aten/src/ATen/quantized/Quantizer.h

```diff
@@ -0,0 +1,226 @@
+ *
+ * When you add new types of leaf Quantizer class, please also
+ * make sure to add a corresponding QScheme enum since
+ * they should have one to one mapping.
+ *
+ * Note about intrusive_ptr:
+ * QTensor holds an intrusive_ptr to Quantizer, and multiple Tensor can
+ * share the same Quantizer.
+ */
+struct CAFFE2_API Quantizer : public c10::intrusive_ptr_target {
+  QScheme qscheme_;
+  Quantizer() {}
+  Quantizer(QScheme qscheme) : qscheme_(qscheme) {}
+  virtual ~Quantizer();
+
+  // Copied from torch/csrc/jit/scope.h
+  QuantizerPtr intrusive_from_this() {
+    c10::raw::intrusive_ptr::incref(this); // we are creating a new pointer
+                                           // from a raw `this` pointer
+                                           // so we need to bump the refcount
+                                           // to account for this ownership
+    return c10::intrusive_ptr<Quantizer>::reclaim(this);
+  }
+
+  virtual QScheme qscheme() {
+    return qscheme_;
+  }
+
+  /**
+   * quantize a float Tensor into a quantized Tensor.
+   */
+  virtual QTensor quantize(RealTensor t) = 0;
+
+  /**
+   * dequantize a quantized Tensor into a float Tensor.
+   */
+  virtual RealTensor dequantize(QTensor t) = 0;
+};
+
+/**
+ * UniformQuantizer is the parent class for all uniform quantizers.
+ * These quantization scheme will map float value uniformly to
+ * the quantized value. For example, affine quantizer is
+ * the most commonly used scheme in this category.
+ */
+struct CAFFE2_API UniformQuantizer : public Quantizer {
+  UniformQuantizer() {}
+  UniformQuantizer(QScheme qscheme): Quantizer(qscheme) {}
+  virtual ~UniformQuantizer();
+};
+
+/**
+ * NonUniformQuantizer is the parent class for all non-uniform quantizers.
+ * These quantization scheme may map float value non-uniformly to the quantized
+ * value. K-means quantization is a representative example in this category.
+ */
+struct CAFFE2_API NonUniformQuantizer : public Quantizer {
+  NonUniformQuantizer() {}
+  NonUniformQuantizer(QScheme qscheme): Quantizer(qscheme) {}
+  virtual ~NonUniformQuantizer();
+
+};
+
+// There is also StochasticQuantizer which is uniform but not affine
+
+/**
+ * AffineQuantizer uses affine transformation to do quantization.
+ *
+ * For quantize:
+ * Y = clamp((X * scale + zero_point, min, max)
+ * For dequantize:
+ * X = (Y - zero_point) / scale
+ */
+struct CAFFE2_API AffineQuantizer : public UniformQuantizer {
+  AffineQuantizer() {}
+  AffineQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~AffineQuantizer();
+};
+
+/**
+ * SymmetricQuantizer is similar to AffineQuantizer except that it
+ * does not have zero_point
+ *
+ * For quantize:
+ * Y = clamp(X * scale, min, max)
+ * For dequantize:
+ * X = Y / scale
+ */
+struct CAFFE2_API SymmetricQuantizer : public UniformQuantizer {
+  SymmetricQuantizer() {}
+  SymmetricQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~SymmetricQuantizer();
```

Comment:
nit: you don't need to put destructors in descendant classes (and even if you did, 'override' instead of 'virtual' should be used)